### PR TITLE
Render Graph Async Compute

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
# Purpose of this PR

Implementation of Async Compute and Pass pruning in Render Graph.
Now enabling the async compute flag on a pass runs the pass asynchronously with the correct dependency barriers.
Also implemented unused pass pruning to remove passes for which the output is never used.

Also includes:
- Moved resource creation/release logic out of the resource registry into the render graph itself.
- Small refactor of ProfilingSamplers to allow explicit Begin/End.
- RenderPass is no longer a subclass of RenderGraph and has been renamed RenderGraphPass
- Added the explicit notion of Transient Texture to the Render Graph to handle temporary textures in a  single pass.
- Improved logging of render graph information.
- Fixed an issue where Empty Exposure texture was not created correctly for render graph.

### CODE REVIEW

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 1]
```diff
@@ -595,65 +586,42 @@ void SetGlobalTextures(RenderGraphContext rgContext, List<TextureHandle> texture
...
-            // If a resource was created for only a single pass, we don't want users to have to declare explicitly the read operation.
-            // So to do that, we also update lastReadIndex on resource writes.
-            // This means that we need to check written resources for destruction too
-            foreach (var resource in writtenTextures)
-            {
-                ref var resourceDesc = ref GetTextureResource(resource);
-                // <= because a texture that is only declared as written in a single pass (and read implicitly in the same pass) will have the default lastReadPassIndex at -1
-                if (!resourceDesc.imported && resourceDesc.lastReadPassIndex <= passIndex)
+                //// This can happen because we release texture in two passes (see ReleaseTexturesForPass) and texture can be present in both passes
+                //if (resourceDesc.rt != null)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs:614]: expect this comment here ?

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs [hunk 2]
```diff
@@ -369,88 +334,363 @@ public RenderGraphBuilder AddRenderPass<PassData>(string passName, out PassData
...
+                    ref CompiledResourceInfo info = ref m_CompiledTextureInfos[texture];
+                    info.producers.Add(passIndex);
+                    passInfo.refCount++;
+
+                    // Writing to an imported texture is considered as a side effect because we don't know what users will do with it outside of render graph.
+                    if (m_Resources.IsTextureImported(texture))
+                        passInfo.hasSideEffect = true;
+                }
+
+                // Can't share the code with a generic func as TextureHandle and ComputeBufferHandle are both struct and can't inherit from a common struct with a shared API (thanks C#)
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs:422]: D:

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs [hunk 3]
```diff
@@ -369,88 +334,363 @@ public RenderGraphBuilder AddRenderPass<PassData>(string passName, out PassData
...
+                    return result;
+            }
+
+            return result;
+        }
+
+
+        void UpdatePassSynchronization(ref CompiledPassInfo currentPassInfo, ref CompiledPassInfo producerPassInfo, int currentPassIndex, int lastProducer, ref int intLastSyncIndex)
+        {
+            // Current pass needs to waoit for pass index lastProducer
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs:502]: small type (waoit)

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs [hunk 4]
```diff
@@ -492,21 +732,61 @@ void PreRenderPassSetRenderTargets(in RenderPass pass, RenderGraphContext rgCont
...
         {
+            RenderGraphPass pass = passInfo.pass;
+
+            if (passInfo.needGraphicsFence)
+                passInfo.fence = rgContext.cmd.CreateAsyncGraphicsFence();
+
+            if (pass.enableAsyncCompute)
+            {
+                // The command buffer has been filled. We can kick the async task.
+                rgContext.renderContext.ExecuteCommandBufferAsync(rgContext.cmd, ComputeQueueType.Background);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs:777]: I am not sure what is happening in Unity behind the scene, but we are losing the option to chose a different queue type here vs non-rendergraph

Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs:777]: Will add ComputeQueueType to the API in the next PR :)

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs (hunk 1 above)
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs:614]: Nope, will remove the code. The test is not necessary anymore.

**Pull request merged**